In [1]:
import whisper
import os
from jiwer import wer

In [ ]:
# Load Whisper model
model_name = "medium.en"
model = whisper.load_model(model_name)

In [2]:
reference_text = "The Rainbow Passage When the sunlight strikes raindrops in the air, they act as a prism and form a rainbow. The rainbow is a division of white light into many beautiful colors. These take the shape of a long round arch, with its path high above, and its two ends apparently beyond the horizon. There is , according to legend, a boiling pot of gold at one end. People look, but no one ever finds it. When a man looks for something beyond his reach, his friends say he is looking for the pot of gold at the end of the rainbow. Throughout the centuries people have explained the rainbow in various ways. Some have accepted it as a miracle without physical explanation. To the Hebrews it was a token that there would be no more universal floods. The Greeks used to imagine that it was a sign from the gods to foretell war or heavy rain. The Norsemen considered the rainbow as a bridge over which the gods passed from earth to their home in the sky. Others have tried to explain the phenomenon physically. Aristotle thought that the rainbow was caused by reflection of the sun's rays by the rain. Since then physicists have found that it is not reflection, but refraction by the raindrops which causes the rainbows. Many complicated ideas about the rainbow have been formed. The difference in the rainbow depends considerably upon the size of the drops, and the width of the colored band increases as the size of the drops increases. The actual primary rainbow observed is said to be the effect of super-imposition of a number of bows. If the red of the second bow falls upon the green of the first, the result is to give a bow with an abnormally wide yellow band, since red and green light when mixed form yellow. This is a very common type of bow, one showing mainly red and yellow, with little or no green or blue."

NH group: 

In [13]:
transcriptions = {}
all_references = []
all_hypotheses = []

audio_dir = os.path.join("CorpusOfDeafSpeech","Normal Hearing")
NH_subject_list = ["NH01","NH02","NH03","NH04","NH05","NH06","NH07","NH08","NH09","NH10"]
passage_name = "RP"
# Iterate over audio files
for (i,NH_subject) in enumerate(NH_subject_list): 
    i = i + 1
    dir_path = os.path.join(audio_dir, NH_subject, "Passages")

    for filename in os.listdir(dir_path):
        if (filename.endswith(".wav") or filename.endswith(".WAV")) and filename.startswith("NH"+str(i)+passage_name):
            file_path = os.path.join(dir_path, filename)
            print(f"Transcribing {filename}...")
            
            # Transcribe the audio file
            result = model.transcribe(file_path, language="en")
            transcription = result["text"]
            transcriptions[filename] = transcription
            
            # Save transcription to file
            filename_stripped = os.path.splitext(filename)[0]  # Get filename without extension
            txt_path = os.path.join(dir_path,filename_stripped+"_medium.en.txt")
            with open(txt_path, "w", encoding="utf-8") as f:
                f.write(transcription)
            # print(f"Transcription saved to {txt_path}")
            
            # Append reference and hypothesis for WER calculation
            all_references.append(reference_text)
            all_hypotheses.append(transcription)
    
    

# Concatenate all references and hypotheses
concatenated_reference = " ".join(all_references)
concatenated_hypothesis = " ".join(all_hypotheses)

# Calculate overall WER
overall_wer_NH = wer(concatenated_reference, concatenated_hypothesis)

Transcribing NH1RPDR2.WAV...
Transcribing NH2RPDR1.wav...
Transcribing NH3RPDR2.WAV...
Transcribing NH4RPDR2.wav...
Transcribing NH5RPDR2.wav...
Transcribing NH6RPDR2.wav...
Transcribing NH7RPDR1.wav...
Transcribing NH8RPDR2.wav...
Transcribing NH9RPDR2.wav...
Transcribing NH10RPDR2.wav...


Load transcribed text and calculate WER: 

In [ ]:
all_references = []
all_hypotheses = []

audio_dir = os.path.join("CorpusOfDeafSpeech","Normal Hearing")
NH_subject_list = ["NH01","NH02","NH03","NH04","NH05","NH06","NH07","NH08","NH09","NH10"]
passage_name = "RP"
# Iterate over audio files
for (i,NH_subject) in enumerate(NH_subject_list): 
    i = i + 1
    dir_path = os.path.join(audio_dir, NH_subject, "Passages")

    for filename in os.listdir(dir_path):
        if (filename.endswith(".wav") or filename.endswith(".WAV")) and filename.startswith("NH"+str(i)+passage_name):
            
            filename_stripped = os.path.splitext(filename)[0]  # Get filename without extension
            txt_path = os.path.join(dir_path,filename_stripped+"_medium.en.txt")
            with open(txt_path, "r", encoding="utf-8") as f:
                transcription = f.read()
            
            # Append reference and hypothesis for WER calculation
            all_references.append(reference_text)
            all_hypotheses.append(transcription)
    

# Concatenate all references and hypotheses
concatenated_reference = " ".join(all_references)
concatenated_hypothesis = " ".join(all_hypotheses)

# Calculate overall WER
overall_wer_NH = wer(concatenated_reference, concatenated_hypothesis)
print(f"WER_NH: {overall_wer_NH}")

WER_NH: 0.05225225225225225


Model size comparison

| model size |  WER   |
| ---------- | ------ |
| tiny       | 11.77% |
| tiny.en    |  9.61% |
| base       |  8.92% |
| base.en    |  7.09% |
| small      |  6.91% |
| small.en   |  5.83% |
| medium     |  5.68% |
| medium.en  |  5.23% |
| large (v3) | 21.86% |
| turbo (v3) | 19.49% |

another visualisaton: 

| model size |  WER (English-only model)  |  WER (multilingual model)  |
| ---------- | -------------------------- | -------------------------- |
| tiny       |  9.61% | 11.77% |
| base       |  7.09% |  8.92% |
| small      |  5.83% |  6.91% |
| medium     |  5.23% |  5.68% |
| large (v3) |   N/A  | 21.86% |
| turbo (v3) |   N/A  | 19.49% |

In [9]:
# show WER of each file trascribed by "turbo" model
for ref,hyp in zip(all_references, all_hypotheses):
    print(wer(ref,hyp))

0.43843843843843844
0.036036036036036036
0.11411411411411411
0.08408408408408409
0.07207207207207207
0.12612612612612611
0.5165165165165165
0.03303303303303303
0.3183183183183183
0.21021021021021022


HI group: 

In [16]:
transcriptions = {}
all_references = []
all_hypotheses = []

audio_dir = os.path.join("CorpusOfDeafSpeech","Deaf") #"CorpusOfDeafSpeech" 
passage_name = "RP"
# Iterate over audio files
for i in range(31):
    i = i + 1
    HI_subject = "Subject "+str(i)
    dir_path = os.path.join(audio_dir, HI_subject, "Passages")
    if i != 11:     # subject 11 doesn't have any passage recordings
        for filename in os.listdir(dir_path):
            if (filename.endswith(".wav") or filename.endswith(".WAV")) and filename.startswith("S"+str(i)+passage_name):
                file_path = os.path.join(dir_path, filename)
                print(f"Transcribing {filename}...")
                
                # Transcribe the audio file
                result = model.transcribe(file_path)
                transcription = result["text"]
                transcriptions[filename] = transcription
                
                # Save transcription to file
                filename_stripped = os.path.splitext(filename)[0]  # Get filename without extension
                txt_path = os.path.join(dir_path,filename_stripped+"_medium.en.txt")
                with open(txt_path, "w", encoding="utf-8") as f:
                    f.write(transcription)
                # print(f"Transcription saved to {txt_path}")
                
                # Append reference and hypothesis for WER calculation
                all_references.append(reference_text)
                all_hypotheses.append(transcription)
    
    

# Concatenate all references and hypotheses
concatenated_reference = " ".join(all_references)
concatenated_hypothesis = " ".join(all_hypotheses)

# Calculate overall WER
overall_wer_HI = wer(concatenated_reference, concatenated_hypothesis)

Transcribing S1RPDR2.wav...
Transcribing S2RPDR2.WAV...
Transcribing S3RPDR2.wav...
Transcribing S4RPDR2.wav...
Transcribing S5RPDR2.WAV...
Transcribing S6RPDR2.WAV...
Transcribing S7RPDR2.wav...
Transcribing S8RPDR2.WAV...
Transcribing S9RPDR2.wav...
Transcribing S10RPDR2.WAV...
Transcribing S12RPDR2.wav...
Transcribing S13RPDR2.WAV...
Transcribing S14RPDR2.wav...
Transcribing S15RPDR2.wav...
Transcribing S16RPDR2.wav...
Transcribing S17RPDR2.wav...
Transcribing S18RPDR2.WAV...
Transcribing S19RPDR2.wav...
Transcribing S20RPDR2.WAV...
Transcribing S21RPDR2.wav...
Transcribing S22RPDR2.wav...
Transcribing S23RPDR2.wav...
Transcribing S24RPDR2.WAV...
Transcribing S25RPDR2.WAV...
Transcribing S26RPDR2.wav...
Transcribing S27RPDR2.wav...
Transcribing S28RPDR2.wav...
Transcribing S29RPDR2.wav...
Transcribing S30RPDR2.wav...
Transcribing S31RPDR2.wav...


In [17]:
print(overall_wer_HI)

0.6927927927927928


In [5]:
## calculate WER for each speech intelligibility level
HI_intell_high = [1,4,5,8,9,10]
HI_intell_med = [6,7,12,15,23,25,26,27,29,30,31]
HI_intell_low = [2,3,13,14,16,17,18,19,20,21,22,24,28]
# import numpy as np
# print(np.sort([*HI_intell_high, *HI_intell_med, *HI_intell_low]))

def wer_from_txt_HI(i_li):
    all_references = []
    all_hypotheses = []
    audio_dir = "CorpusOfDeafSpeech"
    passage_name = "RP"
    # Iterate over audio files
    for i in i_li:          # os.listdir(audio_dir)
        HI_subject = "Subject "+str(i)
        dir_path = os.path.join(audio_dir, HI_subject, "Passages")
        if i != 11:     # subject 11 doesn't have any passage recordings
            for filename in os.listdir(dir_path):
                if (filename.endswith(".wav") or filename.endswith(".WAV")) and filename.startswith("S"+str(i)+passage_name):
                    print(f"Getting transcription for {filename}...")
                    
                    filename_stripped = os.path.splitext(filename)[0]  # Get filename without extension
                    txt_path = os.path.join(dir_path,filename_stripped+"_medium.en.txt")

                    # Read from transcription file
                    with open(txt_path, "r", encoding="utf-8") as f:
                        transcription = f.read()
                    
                    # Append reference and hypothesis for WER calculation
                    all_references.append(reference_text)
                    all_hypotheses.append(transcription)
        
    # Concatenate all references and hypotheses
    concatenated_reference = " ".join(all_references)
    concatenated_hypothesis = " ".join(all_hypotheses)

    # Calculate overall WER
    overall_wer_HI = wer(concatenated_reference, concatenated_hypothesis)
    return overall_wer_HI

WER_high = wer_from_txt_HI(HI_intell_high)
WER_med = wer_from_txt_HI(HI_intell_med)
WER_low = wer_from_txt_HI(HI_intell_low)
print(f"WER for HI speakers with high speech intelligibility: {WER_high}")
print(f"WER for HI speakers medium speech intelligibility: {WER_med}")
print(f"WER for HI speakers low speech intelligibility: {WER_low}")

Getting transcription for S1RPDR2.wav...
Getting transcription for S4RPDR2.wav...
Getting transcription for S5RPDR2.WAV...
Getting transcription for S8RPDR2.WAV...
Getting transcription for S9RPDR2.wav...
Getting transcription for S10RPDR2.WAV...
Getting transcription for S6RPDR2.WAV...
Getting transcription for S7RPDR2.wav...
Getting transcription for S12RPDR2.wav...
Getting transcription for S15RPDR2.wav...
Getting transcription for S23RPDR2.wav...
Getting transcription for S25RPDR2.WAV...
Getting transcription for S26RPDR2.wav...
Getting transcription for S27RPDR2.wav...
Getting transcription for S29RPDR2.wav...
Getting transcription for S30RPDR2.wav...
Getting transcription for S31RPDR2.wav...
Getting transcription for S2RPDR2.WAV...
Getting transcription for S3RPDR2.wav...
Getting transcription for S13RPDR2.WAV...
Getting transcription for S14RPDR2.wav...
Getting transcription for S16RPDR2.wav...
Getting transcription for S17RPDR2.wav...
Getting transcription for S18RPDR2.WAV...
G

NH v.s. HI comparison

| model size | WER_NH | WER_HI |
| ---------- | ------ | ------ |
| medium.en  |  5.23% | 69.28% |


HI WER across speech intelligibility

| Intelligibility | high | medium | low | overall |
| --------------- | ---- | ------ | ----| ------- |
|       WER       |   6.06%     |      35.87%     |     126.54%    |     69.28%    |
|       WERR      |  15.90%     |     586.52%     |    2321.75%    |   1225.86%    |

WER Relative (WERR): 

In [7]:
# overall_wer_HI = 0.6927927927927928
# overall_wer_NH = 0.05225225225225225
print(f"WER_NH: {overall_wer_NH*100:.2f}%")
print(f"WER_HI: {overall_wer_HI*100:.2f}%")

# calculate WERR
werr = (overall_wer_HI - overall_wer_NH) / overall_wer_NH
print(f"WERR: {werr*100:.2f}%")

werr_high = (WER_high - overall_wer_NH) / overall_wer_NH
print(f"WERR (high): {werr_high*100:.2f}%")
werr_med = (WER_med - overall_wer_NH) / overall_wer_NH
print(f"WERR (medium): {werr_med*100:.2f}%")
werr_low = (WER_low - overall_wer_NH) / overall_wer_NH
print(f"WERR (low): {werr_low*100:.2f}%")

WER_NH: 5.23%
WER_HI: 69.28%
WERR: 1225.86%
WERR (high): 15.90%
WERR (medium): 586.52%
WERR (low): 2321.75%
